store/src/java/com/zimbra/cs/event/SolrEventStore.java

```diff
@@ -58,6 +77,158 @@ protected void deleteEventsByDataSource(String dataSourceId) throws ServiceExcep
+            return getQueryToSearchContactAsSenderOrReceiver(contact).toString();
+        }
+    }
+
+    private String getContactFrequencyQueryForTimeRange(ContactAnalytics.ContactFrequencyTimeRange timeRange) throws ServiceException {
+        if(ContactAnalytics.ContactFrequencyTimeRange.FOREVER.equals(timeRange)) {
+            return new MatchAllDocsQuery().toString();
+        }
+        TermRangeQuery rangeQuery = TermRangeQuery.newStringRange(LuceneFields.L_EVENT_TIME, getContactFrequencyQueryStartDate(timeRange), "NOW", true, true);
+        return rangeQuery.toString();
+    }
+
+    private String getContactFrequencyQueryStartDate(ContactAnalytics.ContactFrequencyTimeRange timeRange) throws ServiceException {
+        switch (timeRange) {
+        case LAST_DAY:
+            return "NOW-1DAY";
+        case LAST_WEEK:
+            return "NOW-7DAY";
+        case LAST_MONTH:
+            return "NOW-1MONTH";
+        default:
+            throw ServiceException.INVALID_REQUEST("Time range not supported " + timeRange, null);
+        }
+    }
+
+    @Override
+    public List<ContactFrequencyGraphDataPoint> getContactFrequencyGraph(String contact, ContactAnalytics.ContactFrequencyGraphTimeRange timeRange) throws ServiceException {
+        LocalDateTime startDate = getStartDateForContactFrequencyGraphTimeRange(timeRange);
+        LocalDateTime endDate = LocalDateTime.now();
+        String aggregationBucket = getAggregationBucketForContactFrequencyGraphTimeRange(timeRange);
+
+        SolrQuery solrQuery = new SolrQuery();
+        solrQuery.setQuery(new MatchAllDocsQuery().toString());
+        solrQuery.addDateRangeFacet(LuceneFields.L_EVENT_TIME, Timestamp.valueOf(startDate), Timestamp.valueOf(endDate), aggregationBucket);
+        solrQuery.addFilterQuery(getQueryToSearchContactAsSenderOrReceiver(contact).toString());
+
+        if (solrHelper.needsAccountFilter()) {
+            solrQuery.addFilterQuery(getAccountFilter(accountId));
+        }
+
+        QueryResponse response = (QueryResponse) solrHelper.executeRequest(accountId, solrQuery);
+
+        List<ContactFrequencyGraphDataPoint> graphDataPoints = Collections.emptyList();
+        List<RangeFacet> facetRanges = response.getFacetRanges();
+        if(facetRanges != null && facetRanges.size() > 0) {
+            List<RangeFacet.Count> rangeFacetResult = facetRanges.get(0).getCounts();
+            graphDataPoints = new ArrayList<>(rangeFacetResult.size());
+            for (RangeFacet.Count rangeResult : rangeFacetResult) {
+                graphDataPoints.add(new ContactFrequencyGraphDataPoint(rangeResult.getValue(), rangeResult.getCount()));
+            }
+        }
+        return graphDataPoints;
+    }
+
+    private String getAggregationBucketForContactFrequencyGraphTimeRange(ContactAnalytics.ContactFrequencyGraphTimeRange timeRange) throws ServiceException {
+        switch (timeRange) {
+        case CURRENT_MONTH:
+            return "+1DAY/DAY";
```

Comment:
Don't quite understand what this is.  Perhaps a comment to explain the format and why is has been chosen to be related to the current month?  Maybe similar comments for the other cases too.

Reply:
"+1DAY/DAY" represent the "Gap/Rounding" needed for Solr facet query. I will add comments to the code.